stan/math/prim/mat/fun/log_mix.hpp

```diff
@@ -0,0 +1,124 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_LOG_MIX_HPP
+#define STAN_MATH_PRIM_MAT_FUN_LOG_MIX_HPP
+
+#include <stan/math/prim/scal/meta/partials_return_type.hpp>
+#include <stan/math/prim/scal/meta/operands_and_partials.hpp>
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/err/check_bounded.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/fun/value_of.hpp>
+#include <stan/math/prim/mat/meta/is_constant_struct.hpp>
+#include <stan/math/prim/mat/fun/log_sum_exp.hpp>
+#include <stan/math/prim/mat/fun/log.hpp>
+#include <stan/math/prim/mat/fun/exp.hpp>
+#include <stan/math/prim/mat/fun/max.hpp>
+#include <stan/math/prim/mat/fun/dot_product.hpp>
+#include <stan/math/prim/mat/fun/value_of.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <string>
+#include <algorithm>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Return the log mixture density with specified mixing proportion
+     * and log densities.
+     *
+     * \f[
+     * \frac{\partial }{\partial p_x}
+     * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+     * \exp^{\log\left(p_n\right)+d_n}\right)
+     * = \frac{e^{d_x-a}}{e^{d_1-a}p_1 + \cdot\cdot\cdot+e^{d_m-a}p_m}
+     * \f]
+     *
+     * \f[
+     * \frac{\partial }{\partial d_x}
+     * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+     * \exp^{\log\left(p_n\right)+d_n}\right)
+     *  = \frac{e^{d_x-a}p_x}{e^{d_1-a}p_1 + \cdot\cdot\cdot+e^{d_m-a}p_m}
+     * \f]
+     *
+     * \f[
+     * \mbox{where } a=max(d)
+     * \f]
+     *
+     * @param theta vector of mixing proportions in [0, 1].
+     * @param lambda vector of log densities.
+     * @return log mixture of densities in specified proportion
+     */
+    template <typename T_theta, typename T_lam>
+    typename return_type<T_theta, T_lam>::type
+    log_mix(const T_theta& theta, const T_lam& lambda) {
+      static const char* function = "log_mix";
+      typedef typename stan::partials_return_type<T_theta, T_lam>::type
+        T_partials_return;
+
+      const size_t N = theta.size();
```

Comment:
I'd use https://github.com/stan-dev/math/blob/develop/stan/math/prim/scal/meta/length.hpp here. I think Eigen types use ints and std::vectors use size_ts (so this sorta thing gives compiler warnings sometimes).